common/util/auto_pop_stack.h

```diff
@@ -0,0 +1,117 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_COMMON_UTIL_AUTO_POP_STACK_H_
+#define VERIBLE_COMMON_UTIL_AUTO_POP_STACK_H_
+
+#include <iterator>
+#include <type_traits>
+#include <vector>
+#include "common/util/iterator_adaptors.h"
+
+namespace verible {
+
+// AutoPop stack class. Templated version of SyntaxTreeContext::AutoPop
+template <typename T>
+class AutoPopStack {
+ public:
+  typedef T value_type;
+  typedef AutoPopStack<value_type> this_type;
+
+  typedef value_type& reference_type;
+  typedef const value_type& const_reference_type;
```

Comment:
Let's follow STL convention and call this `const_reference`.

Reply:
Also fixed `reference`